packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+   * @see {@link PermissionController.acceptPermissionsRequest} For the method
+   * that _accepts_ the request and resolves the user approval promise.
+   * @see {@link PermissionController.rejectPermissionsRequest} For the method
+   * that _rejects_ the request and the user approval promise.
+   * @param origin - The origin of the grantee subject.
+   * @param requestedPermissions - The requested permissions.
+   * @param id - The id of the permissions request.
+   * @param preserveExistingPermissions - Whether to preserve the subject's
+   * existing permissions.
+   * @returns The granted permissions and request metadata.
+   */
+  async requestPermissions(
+    origin: OriginString,
+    requestedPermissions: RequestedPermissions,
+    id: string = nanoid(),
+    preserveExistingPermissions = true,
+  ): Promise<
+    [
+      Record<
+        string,
+        ExtractPermission<
+          ControllerPermissionSpecification,
+          ControllerCaveatSpecification
+        >
+      >,
+      { id: string; origin: OriginString },
+    ]
+  > {
+    this.validateRequestedPermissions(origin, requestedPermissions);
+    const metadata: PermissionsRequest['metadata'] = {
+      id,
+      origin,
+    };
+
+    const permissionsRequest: PermissionsRequest = {
+      metadata,
+      permissions: requestedPermissions,
+    };
+
+    const { permissions: approvedPermissions, ...requestData } =
+      await this.requestUserApproval(permissionsRequest);
+    if (Object.keys(approvedPermissions).length === 0) {
+      // The request should already have been rejected if this is the case.
+      throw internalError(
+        `Approved permissions request for origin "${origin}" contains no permissions.`,
+        { requestedPermissions },
+      );
+    }
+
+    return [
+      this.grantPermissions({
+        subject: { origin },
+        approvedPermissions,
+        preserveExistingPermissions,
+        requestData,
+      }),
+      metadata,
+    ];
+  }
+
+  /**
+   * Adds a request to the {@link ApprovalController} using the
+   * {@link AddApprovalRequest} action.
+   *
+   * @param permissionsRequest - The permissions request object.
+   * @returns The approved permissions request object.
+   */
+  private async requestUserApproval(permissionsRequest: PermissionsRequest) {
+    const { origin, id } = permissionsRequest.metadata;
+    return (await this.messagingSystem.call(
+      'ApprovalController:addRequest',
+      {
+        id,
+        origin,
+        requestData: permissionsRequest,
+        type: MethodNames.requestPermissions,
+      },
+      true,
+    )) as PermissionsRequest;
+  }
+
+  /**
+   * Validates requested permissions. Throws if validation fails.
+   *
+   * All properties of {@link RequestedPermissions} values are ignored during
+   * request processing in this controller except `caveats`, which is validated
+   * by {@link PermissionController.computeCaveats}, caveat validators, and
```

Comment:
What was meant by `computeCaveats`? That seems to not exist.

Reply:
This whole description seems confusing to me, after looking at the implementation more closely. All that we're validating here is that the requested permissions are the expected shape (an object), that they're non-empty, and that `methodName` matches each `parentCapability` and is valid accoridng to the spec.

It doesn't validate anything about caveats, and it doesn't call other validators.

Reply:
11e318e